taggit/admin.py

```diff
@@ -14,3 +19,60 @@ class TagAdmin(admin.ModelAdmin):
     ordering = ["name", "slug"]
     search_fields = ["name"]
     prepopulated_fields = {"slug": ["name"]}
+    actions = ["render_tag_form"]
+
+    def get_urls(self):
+        urls = super().get_urls()
+        custom_urls = [
+            path(
+                "merge-tags/",
+                self.admin_site.admin_view(self.merge_tags_view),
+                name="merge_tags",
+            ),
+        ]
+        return custom_urls + urls
+
+    @admin.action(description="Merge selected tags")
+    def render_tag_form(self, request, queryset):
+        selected = request.POST.getlist(admin.helpers.ACTION_CHECKBOX_NAME)
+        if not selected:
+            self.message_user(request, "Please select at least one tag.")
+            return redirect(request.get_full_path())
+
+        selected_tag_ids = ",".join(selected)
+        redirect_url = f"{request.get_full_path()}merge-tags/"
+
+        request.session["selected_tag_ids"] = selected_tag_ids
+
+        return redirect(redirect_url)
+
+    def merge_tags_view(self, request):
+        selected_tag_ids = request.session.get("selected_tag_ids", "").split(",")
+        if request.method == "POST":
+            form = MergeTagsForm(request.POST)
+            if form.is_valid():
+                new_tag_name = form.cleaned_data["new_tag_name"]
+                new_tag, created = Tag.objects.get_or_create(name=new_tag_name)
+                with transaction.atomic():
+                    for tag_id in selected_tag_ids:
+                        tag = Tag.objects.get(id=tag_id)
+                        tagged_items = TaggedItem.objects.filter(tag=tag)
+                        for tagged_item in tagged_items:
+                            tagged_item.tag = new_tag
+                            TaggedItem.objects.filter(tag=tag).update(tag=new_tag)
+                        # tag.delete()  #this will delete the selected tags after merge...leaving out for now
+
+                self.message_user(request, "Tags have been merged")
+                # clear the selected_tag_ids from session after merge is complete
+                request.session.pop("selected_tag_ids", None)
+                return redirect("..")
+            else:
+                print(f"Form errors: {form.errors}")
+                self.message_user(request, "Form is invalid.", level="error")
+
+        context = {
+            "form": MergeTagsForm(),
```

Comment:
So when you're validating a form, you want to make sure you pass that form you tried to validate back into the context.

```
form = MergeTagsForm(data=request.POST  if request.method == "POST" else None)
if request.method == "POST":
  if form.is_valid():
   ....
...
context = {
  "form": form, 
   ...
}
```
if you do that the form errors will be properly displayed if there are any

Reply:
This isn't working because the context is defined outside of the conditional where the form is defined. I assumed the form that got passed in from `MergeTagsForm()` would work. It renders the form and that's how the max_length validation is implemented.